### ISSUE TITLE
Preferences->Windows->"Prevent focus stealing" has no effect.

### ISSUE THREAD

**Issue by crobarcro** [linuxmint/Cinnamon#8059]
```
 * Cinnamon version (cinnamon --version)
 * Distribution - (Mint 17.2, Arch, Fedora 25, etc...)
 * Graphics hardware *and* driver used
 * 32 or 64 bit
 * Attach /home/<username>/.xsession-errors, or /var/log/syslog
 ```

**Issue**
The setting Preferences->Windows->"Prevent focus stealing" has no effect. When I activate it, it does not prevent focus stealing.

**Steps to reproduce**

Navigate to Preferences->Windows and activate the "Prevent focus stealing" options as in the screenshot below.

![screenshot from 2018-11-22 10-03-17](https://user-images.githubusercontent.com/3159692/48935336-45356b00-eeff-11e8-94f8-93f983e4f03d.png)

Luahcn a program, e.g. Firefox web browser, the newly launched program will grab focus (also has no effect after logout/login or full reboot).

**Expected behaviour**

Programs will be prevented from stealing focus.

**Other information**

This is on the official release of Mint Linux 19 Cinnamon, with Cinnamon version 3.8.9.

**Comment by icarter09** [comment]
@crobarcro did the merged PR fix this issue?

**Comment by densha** [comment]
@icarter09 @crobarcro Is there any update on this? This is still an issue with the most recent version of Linux Mint. I enabled the setting and apps still steal focus, even if the app is in a different workspace.

**Comment by icarter09** [comment]
@densha thanks for the update that this is still an issue. Sounds like the merged PR did not provide the fix for it.

**Comment by TheSHEEEP** [comment]
Indeed still an issue.

It is quite annoying to type something in a window, only to have the focus stolen by some other window that just opened.
KDE Plasma does this quite nicely.

**Comment by devgs** [comment]
Come on, guys! How hard it is to *disable* a feature?

**Comment by FibroJedi** [comment]
Just to help keep this issue live, yes it's **still** an issue. I share @Troi's concern. While normally it's just annoying, when logging into apps, it nearly always happens because I multitask, so yes, often happens when inputting usernames/passwords so is a real security concern.

**Comment by Viissataa** [comment]
STILL an issue! How fucking hard is it?
This is an exceedingly annoying issue, and affects not only cinnamon, but xfce too. This needs attention!

**Comment by mtwebster** [comment]
Does this involve only a single workspace typically? Or only with windows on different workspaces?

Can you guys provide program names to reproduce this please. (The name of the program that steals focus).

**Comment by devgs** [comment]
@mtwebster Consider Telegram client, as an example (but the list is bottomless). Whenever someone texts you something, client's window draws attention and instantly pops up to the front. When it's on a different workspace, view is automatically switched to that workspace.

Worse still, if at that moment you had some app in fullscreen, there's a chance that, when you switch to the original workspace, that app's window would be shifted by a half of a screen up and to the right, and will become unresponsive: you can't focus it again, only kill. It has something to do with post-effects or something. I'm judging from how it looks: half blurred. Unfortunately, I don't have a screenshot to show.

**Comment by mtwebster** [comment]
With telegram, what behavior are you after that isn't provided by turning off 'draw attention to the window' in telegram's notification settings?

I'm not saying that's the 'fix' here, It was just an observation and I'm curious. Are there other programs that exhibit this also that *don't* allow disabling the behavior within their own settings?

If I have that enabled, then it does indeed end up ignoring Cinnamon's setting, but with it disabled it behaves as it should (though ideally it should be popping a temporary item in the window list applet that's highlighted and you can click on to raise the window or switch to its workspace).

**Comment by devgs** [comment]
@mtwebster With notifications disabled, Telegram app simply doesn't emit them. It's like to cut your hand to treat a broken finger. All other app that implement a 'draw attention' action. Another example that comes to mind: Qt Creator, when project build is finished. There are many more apps that I've come across over the years, that I can't specifically remember. Usually, anything that pops a dialog box, while the window is not active, will draw attention and repeat the issue. So, I mean, it's **not** the apps fault but the **system's**.

**Comment by crobarcro** [comment]
> Does this involve only a single workspace typically? Or only with windows on different workspaces?
> 
> Can you guys provide program names to reproduce this please. (The name of the program that steals focus).

Matlab in my case,  if I start loading it, i.e. double click on it#s icon on the desktop, then go do something else, it steals focus when it loads. I'm pretty sure every app does this, and not just when loading. This is on a single workspace, probably also across workspaces (but in this case if you start loading on one workspace, then switch to another, the program ultimately ends up on the current workspace, which I would also prefer to work differrenty actually, but it's a minor issue).

**Comment by devgs** [comment]
So, ultimately, my expectation as a user is: when "Prevent focus stealing" is **on**: DE should do just that. Whenever any one app (window) wants to draw attention, instead of giving it an immediate focus, DE should only allow the app (window) to start blinking in the **window list** panel. 

Otherwise that's a huge security issue: I personally had dozens of times when I type some sensitive data (eg password) in one app and then Telegram pops up and grabs all thos key presses and consumes them as if I was typing there. Only by a sheer luck have I avoided sending it to the telegram chat.

**Comment by NikoKrause** [comment]
> Consider Telegram client, as an example (but the list is bottomless).

With a bottomless list one would think that you could come up with more than one example. 😁

**Comment by devgs** [comment]
@NikoKrause I did. You'd see if you'd read more of my replies. As a general rule: any app that that pops dialog box while being minimized or on another workspace. This makes the list bottomless. Probably, because dialog box is interpreted by DE as the need for attention.

**Comment by Viissataa** [comment]
I have had this happen with JetBrains PyCharm and TI Code composer as
recurring offenders.

Besides security issues, I've had probably 20 compilation errors or BUGS
because there is something like:

for(int i=0; i<N;i++){  cd ../projectir

[rest of the code]
}

C luckily has compile stage but with Python this is insanely annoying and
dangerous!
It needs to get fixed!

I already attempted using windows virtual machine bc of this.

The only sensible mechanism is that focus is absolutely solely granted by
user action. No sane person would ever want anything else, or they should
be forced to fork a fucking kernel to do something like that.

**Comment by Viissataa** [comment]
I already looked around and concerningly "[DE name] focus steal prevention
doesn't work" gets hits with xfce, kde, and cinnamon at least.

Why in devils name is there such a thing as "focus stealing prevention"
anyway?

Focus stealing should be some exotic wild feature that DEs disable by
themselves every 10 minutes without asking if the user ever manages to get
it enabled.

**Comment by mtwebster** [comment]
@Viissataa can you provide more specific details about pycharm stealing focus, such as when and possibly how to reproduce it, or at least the type of scenario that reproduces it.

I've found some issues that I'm fixing, which resolves this for telegram and should fix any other *well-behaved* program.

Matlab is not one of those - no matter what, it steals focus when it wants. This seems like a very common complaint and not unique to cinnamon. There's nothing I can do about it on our end without probably breaking everything else - this isn't a simple issue, though it may seem so.

I did test this:
https://github.com/joka90/matlab-focus-fixer
and can confirm it works (at least with my changes applied on our end).

I have a couple more programs I'd like to test before I push the changes I have.

**Comment by Viissataa** [comment]
I have installed PyCharm by downloading the newest deb package from
JetBrains site.

To reproduce:
1: First open PyCharm, create a project, and close PyCharm.

2: open a console/terminal.

3: Start PyCharm from a shortcut of choise (desktop, toolbar, etc)

4: Focus on terminal and wait.

PyCharm takes some time to load assets, and will soon steal focus from
terminal.

**Comment by FibroJedi** [comment]
Just caught up here so I ran my own test

1. As above, load terminal, type something
2. Open a document in Libre Office Writer 
3. Re-focus on Terminal and type something
4. The Libre Office document loads, and steals the focus.

Hope this helps.

Earlier someone did say "the list is bottomless" - while I can't attest to that, I do know it happens a LOT, when you load something and go back to what you were doing until it's loaded.

**Comment by Viissataa** [comment]
My suggestion for a feature is to completely ignore any and all focus grab
requests not originating from user input, without a solitary exception, or
provide any way for application api to circumvent this.

**Comment by FibroJedi** [comment]
The one I don't mind is when Updates loads and requests authentication, but that's the only excusable one I think.

**Comment by devgs** [comment]
My current understanding is: X11 protocol is so lax that it allows client to do anything. All clients are fully trusted to behave sanely. That's probably explains such a bizarrely common issue and the one that is so hard to fix. And a ton of more issues, such as some apps causing Xorg to spin on 100% CPU, by simply spamming it with useless events.

Unless it's properly fixed in Xorg, all DE can do is to "apply medical tape to a leaking boat".

Therefore, @mtwebster, my gratitude for taking your time and trying!

**Comment by mtwebster** [comment]
@devgs is correct (I've been attempting to figure out how to express these realities without being crucified here - thanks :)

I believe I've fixed issues related to focus-stealing by already-running programs such as telegram. Instead of stealing focus, they should merely notify by flashing in the window list. The 'prevent windows..' setting is removed. The 'bring windows which require attention...' setting is still there and functions appropriately.

The [focus-fixer](https://github.com/joka90/matlab-focus-fixer) I mentioned seems to function well for more than just matlab (it fixed pycharm also for me). Maybe we'll look into that more later to make it simpler to use in Cinnamon. This may not work without the changes I just made, I'll try to test later without them.

There is a potential for regressions here, but I can address them as they come, and hopefully in a more appropriate manner.

I'll leave this issue open for now.  Any one wishing to test these changes on Mint can use the daily builds PPA (https://launchpad.net/~linuxmint-daily-build-team/+archive/ubuntu/daily-builds). Anyone else needs to build both cinnamon and muffin to test.

Thanks

**Comment by devgs** [comment]
@mtwebster Thanks a lot! 

I'm gonna be testing PPA as soon as possible.

**Comment by Viissataa** [comment]
Thank you for spending time on this.

Interesting, a patch at X11 level would then fix this in all DEs. Sounds
worth an attempt. Maybe I'll try that.

**Comment by devgs** [comment]
@mtwebster kudos! Finally, for both Telegram and Qt Creator the issue is resolved!

Hopefully, the next stable release of **cinnamon** will contain the fix.

**Comment by Moini** [comment]
Just came here because I have this problem on Xfce. Will that require a different fix, or is this a general fix for Linux Mint?

**Comment by FibroJedi** [comment]
Also do we have a vague timeframe on the next stable version of Cinnamon?

**Comment by invernosantigos** [comment]
> Replying to Mtwebster-san, a very practical example of stealing focus in Mint 20.3, is when you open a download box, or an image (with nomacs -- with gthumb or pix it doesn't lose focus) and then just move the mouse to one of the corners of the screen, mostly towards the top (large sensitivity area) and the application window ( nomacs or a dialog box ) immediately goes out of focus. I don't know why, but it happens more with some apps than others. It hardly happens with the ones pre-installed in mint; but on user-installed ones, especially slower ones or those that use a lot of bash, it's fatal. Which suggests that it is more a problem of integration than of focus stealing.

**Comment by Gaturinho** [comment]
> invernosantigos

And another example, this very irritating one, is when you open a box asking for a password. To write the password in the box field, it is necessary to keep the mouse pointer exactly over the box -- if even a millimeter is out of it, it is not possible to enter the text in the field, because the insertion point is only maintained with the mouse strictly on the box, and worse, just a notification or any window pops up asking for focus, that the insertion point has died, and we have to start again, and sometimes, again, and again....

**Comment by elfein727** [comment]
This really sucks because some programs open multiple short-lived loading bars, and typing will sometimes abort those loadings.

**Comment by Moini** [comment]
Yeah, the update manager is the worst of all for me. It interrupts twice for showing a progress bar.

**Comment by elfein727** [comment]
It appears the new focus mode option is either "Smart" or "Strict":

![image](https://user-images.githubusercontent.com/45273247/179370348-7f1f207e-dab3-4c3f-aa11-8ffe6fe4da6f.png)

Neither of those options is what the user wants. The user wants new windows that require focus to .... not steal focus. If anything, the user wants a notification saying that a window requires attention, with options to configure how those notifications play out. Interrupting user input can lead to unpredictable behavior across different windows, so this is a security issue as well.

**Comment by elfein727** [comment]
OOF, just rea the whole comment chain. I see now what is wrong. Damn it, X.... You've damned us again.

**Comment by FibroJedi** [comment]
Ah yes, of course, sorry. Blame the heat wave affecting my brain or some
other poor excuse for rushing my reply!

**Comment by NickNackGus** [comment]
Nearly five years later, and I'm still having this issue - I've had to disable notifications in several apps and websites to avoid switching windows when I don't want to. I get that xorg doesn't seem to respond in a way that makes sense under the hood, but Mate doesn't have this issue, so there's got to be some sort of workaround.

**Comment by FibroJedi** [comment]
Yes, I'm still having the issue also. I often load programs/apps behind what I'm working on, but if that thing has a form element on it (like a login screen), it steals the focus every time, for me.

**Comment by Lanchon** [comment]
the "Prevent focus stealing" option no longer exists in settings. should we close this issue or rename it?

focus stealing is still an issue in 22.1. is a solution planned only for wayland? or should keep hoping for something in x11?

**Comment by Moini** [comment]
Agreed. Lately even notifications (which shouldn't do anything) steal focus, which is super annoying.

**Comment by skorpion9** [comment]
Hi!
This issue is still present, and I (we) ask you to put back this switch, therefore we can choose if a new window will land in the front or in the back of existing windows. This would be very handy, as this focus stealing behavior is very annoying to us.

Example: I open a heavy program, it takes 10-30 s to load. In the meantime I start to do other things, then realize half of my sentence I wrote is missing, due to loss of focus. Not nice.

I hope this is not a big deal, since it was already there. Please bring it back.

**Comment by AFulgens** [comment]
It's crazy for me that since the very first GUI-based OSs this has been an issue and no OS, including *nix distributions have a solution for this, which even for a programmer seems to be "solvable". So... why isn't it? 😅 

The question from Lanchon is the correct one: Are we expecting a solution for this? Are we expecting a Wayland-only solution for this?
